### PR TITLE
Enhance reproducibility documentation

### PR DESCRIPTION
**Pull request by duncanriach**
Improves explanation of non-determinism when running on GPUs. Adds info about `torch.nn.BCELoss` operating non-deterministically on GPUs.

### CODE REVIEW

**Comment by ngimel** [comment]
cc @kurtamohler re: non-determinism of BCE loss. I don't see any atomics operations or any non-determinism source in the current implementation. Was there  non-determinism before the ATen port? Is it non-deterministic now?
@duncanriach do we have a small reproducer for BCE loss non-determinism?

**Comment by kurtamohler** [comment]
The only thing that I can think of at the moment that might be non-deterministic in the cuda version of BCELoss is the reduction, `at::sum()` or `at::mean()`. But I don't know for sure if those are non-deterministic. If there is a reproducer, we could test whether setting `reduction=None` makes it deterministic.

**Comment by duncanriach** [comment]
@ngimel and @kurtamohler, I could not find the use of atomic operations either. I can get back to you with a reproducer.

**Comment by ngimel** [comment]
sum and mean are deterministic on cpu and gpu.

**Comment by duncanriach** [comment]
@ngimel and @kurtamohler, I think I've made a mistake regarding BCE Loss. I don't think that it's actually the source of non-determinism that I'm seeing. Would you still like to take the other suggested modifications in the pull-request?

Also, while you're here, it now appears that non-determinism is being introduced by `torch.repeat_interleave`. Is it possible that the back-prop reduction through the broadcast is using `atomicAdd`? I'm seeing non-determinism when `repeats` is greater than 4 when repeating into the innermost dimension and I'm seeing non-determinism when `repeats` is greater than 16 when repeating into other dimensions.

**Comment by ngimel** [comment]
The other suggested modifications look good. You might want to add `interpolate` to the list of functions with non-deterministic backward. 
Do you have a reproducer for repeat_interleave? I don't know how it is implemented, it is possible that there is some non-determinism there.

**Comment by duncanriach** [comment]
Thanks. I'll add a commit to this PR for the documentation and continue simplifying a reproducer.

**Comment by duncanriach** [comment]
Thanks @ngimel. I've updated and rebased this. Hopefully it's now ready to be merged.

**Comment by ngimel** [comment]
Thanks!

**Review by ngimel — APPROVED** [review]

**Comment by yottabytt** [comment]
@duncanriach @ngimel Does `gather()` or `cumsum()` internally use any of the non-deterministic functions for its forward or backward?

**Comment by ngimel** [comment]
`gather()` does (backward is scatter_add), `cumsum` does not. `gather` is mentioned in reproducibility documentation
